File: libfrag/python/examples/mbe_demo.py
#!/usr/bin/env python3
"""
Many-Body Expansion (MBE) Example for libfrag

This example demonstrates how to use the MBE functionality in libfrag
to perform fragment-based quantum chemistry calculations.
"""

import libfrag
from libfrag.mbe import *
import numpy as np

def main():
    print("=== libfrag Many-Body Expansion Demo ===")
    
    # Create a simple water dimer system
    atoms = [
        libfrag.Atom("O", 0.0, 0.0, 0.0),      # Water 1
        libfrag.Atom("H", 0.76, 0.59, 0.0),
        libfrag.Atom("H", -0.76, 0.59, 0.0),
        libfrag.Atom("O", 3.0, 0.0, 0.0),      # Water 2  
        libfrag.Atom("H", 3.76, 0.59, 0.0),
        libfrag.Atom("H", 2.24, 0.59, 0.0)
    ]
    
    molecule = libfrag.Molecule(atoms)
    print(f"Created water dimer with {len(molecule)} atoms")
    
    # Configure MBE calculation
    config = MBEConfig.default_2body()
    config.set_qm_method("HF")
    config.set_basis_set("6-31G*")
    config.set_fragmentation_scheme(FragmentationScheme.MOLECULAR)
    config.set_distance_cutoff(8.0)
    
    print("\nMBE Configuration:")
    print(config.to_string())
    
    # Test fragment generation
    print("\n=== Fragment Generation ===")
    generator = MBEFragmentGenerator(config)
    
    # Estimate number of combinations
    n_fragments_estimate = 2  # Two water molecules
    for order in range(1, config.max_order() + 1):
        n_combinations = generator.estimate_combinations(n_fragments_estimate, order)
        print(f"{order}-body combinations: {n_combinations}")
    
    # Test MBE results storage
    print("\n=== Results Testing ===")
    results = MBEResults(config)
    
    # Add some mock fragment results
    for i in range(2):  # 1-body terms
        fragment_result = FragmentCalculationResult([i], 1)
        fragment_result.total_energy = -76.0 + i * 0.001  # Mock energies
        fragment_result.qm_method = "HF"
        fragment_result.basis_set = "6-31G*"
        fragment_result.converged = True
        results.add_fragment_result(fragment_result)
    
    # Add 2-body interaction
    fragment_result = FragmentCalculationResult([0, 1], 2)
    fragment_result.total_energy = -152.005  # Mock interaction energy
    fragment_result.qm_method = "HF"
    fragment_result.basis_set = "6-31G*"
    fragment_result.converged = True
    results.add_fragment_result(fragment_result)
    
    print(f"Added {results.n_calculations()} fragment calculations")
    print(f"Total MBE energy: {results.total_energy():.6f} Hartree")
    
    # Print energy decomposition
    print("\nEnergy Contributions:")
    for order in range(1, results.max_order() + 1):
        contribution = results.energy_contribution(order)
        print(f"  {order}-body: {contribution:.6f} Hartree")
    
    # Export results
    print("\n=== Results Export ===")
    print("Summary Report:")
    print(results.summary_report())
    
    print("\nJSON Export (first 200 chars):")
    json_output = results.to_json()
    print(json_output[:200] + "..." if len(json_output) > 200 else json_output)
    
    # Test different configurations
    print("\n=== Configuration Examples ===")
    
    # 3-body configuration
    config_3body = MBEConfig.default_3body()
    print(f"3-body config: max_order={config_3body.max_order()}")
    
    # Custom fragmentation
    custom_config = MBEConfig(2, FragmentationScheme.CUSTOM)
    custom_fragments = [
        [0, 1, 2],  # First water
        [3, 4, 5]   # Second water  
    ]
    custom_config.set_custom_fragments(custom_fragments)
    custom_config.set_qm_method("B3LYP")
    custom_config.set_basis_set("def2-TZVP")
    
    print(f"Custom config: {len(custom_config.custom_fragments())} fragments defined")
    print(f"QM Method: {custom_config.qm_method()}/{custom_config.basis_set()}")
    
    # Test utility functions
    print("\n=== Utility Functions ===")
    n_combinations_2of5 = binomial_coefficient(5, 2)
    print(f"Binomial coefficient C(5,2) = {n_combinations_2of5}")
    
    combinations_3of4 = generate_combinations(4, 3)
    print(f"All 3-combinations from 4 items: {combinations_3of4}")
    
    # Test overlap checking
    combo1 = [0, 1, 2]
    combo2 = [2, 3, 4]
    overlap = combinations_overlap(combo1, combo2)
    print(f"Combinations {combo1} and {combo2} overlap: {overlap}")
    
    print("\n=== Demo Complete ===")
    print("MBE functionality successfully demonstrated!")
    print("\nKey features showcased:")
    print("- Configuration management with different fragmentation schemes")
    print("- Results storage and energy decomposition analysis")
    print("- Fragment combination generation and utilities")
    print("- Export capabilities (JSON, CSV, summary reports)")
    print("- Validation and error checking")
    
    print("\nNote: This demo uses mock data since QM calculator integration")
    print("requires connection to actual quantum chemistry software.")

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"Error: {e}")
        print("Note: This example requires the libfrag Python module to be built and installed.")
        print("Run this after successful compilation of the C++ library with Python bindings.")

File: libfrag/examples/mbe_demo.cpp
#include "libfrag/libfrag.hpp"
#include <iostream>
#include <memory>

using namespace libfrag;

// Example QM Calculator implementation (placeholder)
class ExampleQMCalculator : public QMCalculatorInterface {
public:
    FragmentCalculationResult calculate_energy(
        const Fragment& fragment,
        const std::string& method,
        const std::string& basis_set,
        int charge = 0,
        int multiplicity = 1,
        const std::unordered_map<std::string, std::string>& options = {}) override {
        
        // Create a mock result for demonstration
        FragmentCalculationResult result;
        result.fragment_id = fragment.fragment_id();
        result.n_body_order = 1;  // This would be determined from the fragment
        result.total_energy = -1.0 + static_cast<double>(rand()) / RAND_MAX * 0.1;  // Mock energy
        result.qm_method = method;
        result.basis_set = basis_set;
        result.converged = true;
        result.computation_time = std::chrono::duration<double>(1.0 + static_cast<double>(rand()) / RAND_MAX);
        result.status = "completed";
        
        return result;
    }
    
    bool is_available() const override {
        return true;  // Mock availability
    }
    
    std::string software_info() const override {
        return "ExampleQM v1.0";
    }
};

int main() {
    std::cout << "=== Many-Body Expansion Demo ===" << std::endl;
    
    try {
        // Create a simple test molecule (water dimer)
        std::vector<Atom> atoms = {
            Atom("O", 0.0, 0.0, 0.0),      // Water 1
            Atom("H", 0.76, 0.59, 0.0),
            Atom("H", -0.76, 0.59, 0.0),
            Atom("O", 3.0, 0.0, 0.0),      // Water 2  
            Atom("H", 3.76, 0.59, 0.0),
            Atom("H", 2.24, 0.59, 0.0)
        };
        
        Molecule water_dimer(atoms);
        std::cout << "Created water dimer with " << water_dimer.size() << " atoms" << std::endl;
        
        // Configure MBE calculation
        auto config = MBEConfig::default_2body();
        config.set_qm_method("HF");
        config.set_basis_set("6-31G*");
        config.set_fragmentation_scheme(MBEConfig::FragmentationScheme::MOLECULAR);
        
        std::cout << "\nMBE Configuration:" << std::endl;
        std::cout << config.to_string() << std::endl;
        
        // Create MBE calculator with mock QM interface
        auto qm_calculator = std::make_unique<ExampleQMCalculator>();
        MBECalculator mbe_calculator(std::move(qm_calculator), config);
        
        // Set up progress callback
        mbe_calculator.set_progress_callback([](double progress, const std::string& status) {
            std::cout << "[" << std::fixed << std::setprecision(1) << progress << "%] " << status << std::endl;
        });
        
        // Set up logging callback
        mbe_calculator.set_log_callback([](const std::string& message) {
            std::cout << "[LOG] " << message << std::endl;
        });
        
        // Preview the calculation
        std::cout << "\n=== Calculation Preview ===" << std::endl;
        auto cost_estimate = mbe_calculator.estimate_cost(water_dimer, config);
        for (const auto& [metric, value] : cost_estimate) {
            std::cout << metric << ": " << value << std::endl;
        }
        
        auto fragment_preview = mbe_calculator.preview_fragments(water_dimer, config);
        std::cout << "\nFragment combinations by order:" << std::endl;
        for (const auto& [order, count] : fragment_preview) {
            std::cout << "  " << order << "-body: " << count << " combinations" << std::endl;
        }
        
        // Validate setup
        std::cout << "\n=== Setup Validation ===" << std::endl;
        auto validation_errors = mbe_calculator.validate_setup(water_dimer, config);
        if (validation_errors.empty()) {
            std::cout << "Setup validation passed" << std::endl;
        } else {
            std::cout << "Validation errors:" << std::endl;
            for (const auto& error : validation_errors) {
                std::cout << "  " << error << std::endl;
            }
            return 1;
        }
        
        // Perform MBE calculation (this would fail with current placeholder implementation)
        std::cout << "\n=== Starting MBE Calculation ===" << std::endl;
        std::cout << "NOTE: This is a demo with placeholder implementations." << std::endl;
        std::cout << "The actual calculation would require full implementation of:" << std::endl;
        std::cout << "- Fragment generation and manipulation" << std::endl;
        std::cout << "- QM calculator interface connections" << std::endl;
        std::cout << "- Energy aggregation and analysis" << std::endl;
        
        /* Commented out since it would fail with current placeholder implementation
        try {
            auto results = mbe_calculator.calculate(water_dimer);
            
            std::cout << "\n=== MBE Results ===" << std::endl;
            std::cout << results.summary_report() << std::endl;
            
            std::cout << "\nEnergy Decomposition:" << std::endl;
            std::cout << results.energy_decomposition_table() << std::endl;
            
            // Export results
            std::cout << "\nResults as JSON:" << std::endl;
            std::cout << results.to_json() << std::endl;
            
        } catch (const std::exception& e) {
            std::cout << "Calculation failed (expected with current implementation): " << e.what() << std::endl;
        }
        */
        
        // Demonstrate configuration options
        std::cout << "\n=== Configuration Examples ===" << std::endl;
        
        auto config_3body = MBEConfig::default_3body();
        std::cout << "3-body configuration:" << std::endl;
        std::cout << config_3body.to_string() << std::endl;
        
        // Custom fragmentation example
        MBEConfig custom_config(2, MBEConfig::FragmentationScheme::CUSTOM);
        std::vector<std::vector<std::size_t>> custom_fragments = {
            {0, 1, 2},  // First water molecule
            {3, 4, 5}   // Second water molecule
        };
        custom_config.set_custom_fragments(custom_fragments);
        custom_config.set_qm_method("B3LYP");
        custom_config.set_basis_set("def2-TZVP");
        
        std::cout << "\nCustom fragmentation configuration:" << std::endl;
        std::cout << custom_config.to_string() << std::endl;
        
        std::cout << "\n=== Demo Complete ===" << std::endl;
        std::cout << "The MBE scaffolding has been successfully created!" << std::endl;
        std::cout << "Key components implemented:" << std::endl;
        std::cout << "- MBEConfig: Configuration management" << std::endl;
        std::cout << "- MBEResults: Results storage and analysis" << std::endl;
        std::cout << "- MBEFragmentGenerator: Fragment combination generation" << std::endl;
        std::cout << "- MBECalculator: Main calculation orchestration" << std::endl;
        std::cout << "- QMCalculatorInterface: Abstraction for QM software integration" << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}
